### ISSUE TITLE
Dadapt Lion feedback

### ISSUE THREAD

**Issue by Goldenkoron** [facebookresearch/dadaptation#35]
Hi, I tested Dadapt Lion and Adam for Stable Diffusion training and found that adam version works great and I feel like my models are converging faster than the learning rates I used with ordinary AdamW or Lion. However, the Dadapt Lion is not working for me and my models turn into pure noise within a few hundred steps and complete black shortly after. This also happens at 0.10 learning rate but a little bit slower. Perhaps it'd work in the E-3 or E-4 LR range, I am uncertain.

Update: The model going NaN was my own configuration error, accidentally having it using decouple like dadapt_adam. However with that fixed, the training does not learn at all, even with a significant number of steps it feels the weights do not change at all.

**Comment by adefazio** [comment]
Thanks for the feedback! This is very useful. I am also seeing that the Lion variant is not as effective as Adam. Since Lion is not backed by theory, it's hard to say how to apply D-Adaptation to it. Did you try Lion without D-Adaptation?

**Comment by Goldenkoron** [comment]
I have used normal lion before and was able to train models on the same
dataset in about half the number of epochs as I did on normal AdamW.

**Comment by konstmish** [comment]
I think this might be because the D-Adapted Lion uses `update` in the lines
```
numerator_acum += dlr * torch.dot(update.flatten(), s.flatten()).item()
s.mul_(sqrt_beta2).add_(update, alpha=(1-sqrt_beta2)*dlr)
```
To me, it would make more sense to `grad` instead of `update` since the gradient shows up in all theoretical derivations. In particular, D-Adapted Adam also uses `grad` rather than `exp_avg` or `exp_avg.div_(denom)`.  
@Goldenkoron can you try changing these lines and see if it works?

**Comment by Goldenkoron** [comment]
Okay, suddenly dadapt lion is working for me and I don't really understand why. I ran an a/b test with dadapt_adam and dadapt lion not only fitted my subject better and faster, it suffered less "forgetting" of original model concepts in stable diffusion.

@konstmish I tried your fix and it actually learned slower than unaltered dadapt lion, I really don't know at the moment what I did to fix it. I did reinstall some things since the last time I tried using dadapt lion.

dadapt lion
![image](https://github.com/facebookresearch/dadaptation/assets/112432266/08df2b7f-4506-4fbc-9873-847d59af7f2f)

dadapt adam
![image](https://github.com/facebookresearch/dadaptation/assets/112432266/737a3566-9ecc-4d6c-8bad-b2eefd32da13)

**Comment by konstmish** [comment]
Thanks for checking. It's too bad there is no theory for Lion that would allow us to check the correct way of estimating the stepsize, I suppose we can leave the implementation as is for now.

**Comment by lu-minous** [comment]
> I think this might be because the D-Adapted Lion uses `update` in the lines
> 
> ```
> numerator_acum += dlr * torch.dot(update.flatten(), s.flatten()).item()
> s.mul_(sqrt_beta2).add_(update, alpha=(1-sqrt_beta2)*dlr)
> ```
> 
> To me, it would make more sense to `grad` instead of `update` since the gradient shows up in all theoretical derivations. In particular, D-Adapted Adam also uses `grad` rather than `exp_avg` or `exp_avg.div_(denom)`. @Goldenkoron can you try changing these lines and see if it works?

`update` for numerator_acum, and `grad` for s, then `d`  converge to 0.007, much larger than normal lr(adout 0.0005)
